Remotion/Data/DomainObjects/Mapping/InlineTypeDefinitionWalker.cs

```diff
@@ -0,0 +1,176 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Mapping
+{
+  /// <summary>
+  /// Walks through the given <see cref="TypeDefinition"/>'s hierarchy and executes a handler based on the entity's type.
+  /// </summary>
+  public static class InlineTypeDefinitionWalker
+  {
+    private class InlineTypeDefinitionWalkerImplementation : TypeDefinitionWalker
+    {
+      private readonly Action<ClassDefinition> _classDefinitionHandler;
+      private readonly Action<InterfaceDefinition> _interfaceDefinitionHandler;
+
+      public InlineTypeDefinitionWalkerImplementation (
+          TypeDefinitionWalkerDirection direction,
+          Action<ClassDefinition> classDefinitionHandler,
+          Action<InterfaceDefinition> interfaceDefinitionHandler)
+          : base(direction)
+      {
+        ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+        ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+        _classDefinitionHandler = classDefinitionHandler;
+        _interfaceDefinitionHandler = interfaceDefinitionHandler;
+      }
+
+      /// <inheritdoc />
+      public override void VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        ArgumentUtility.CheckNotNull("classDefinition", classDefinition);
+
+        _classDefinitionHandler(classDefinition);
```

Comment:
I'm questioning if handler should be before or after the base call. Possibly, it depends on the direction of the walk.

This might be extremely usecase dependent. For instance, if we build a cache of all the properties defined by the hierarchy, we want to calculate the base-type first, then our own type.

on the other hand, if we want to return a value once we find it, we typically can stop once the first type returns the value, including our own.

Reply:
It would be possible to add an enum to indicate when the handler is called if that behavior is needed. But I don't think we need it. All of the use cases right now do not require this behavior. In cases where things need to be built, like the property cache, you mentioned, a manual approach is probably better anyway.